src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -404,15 +411,27 @@ private BoundIndexerAccess BindIndexerDefaultArgumentsAndParamsCollection(BoundI
         /// method returns a BoundBadExpression node. The method returns the original
         /// expression without generating any error if the expression has errors.
         /// </summary>
-        private BoundExpression CheckValue(BoundExpression expr, BindValueKind valueKind, BindingDiagnosticBag diagnostics)
+        private BoundExpression CheckValue(BoundExpression expr, BindValueKind valueKind, BindingDiagnosticBag diagnostics
+#if DEBUG
+            , bool dynamificationOfAssignmentResultIsHandled = false
```

Comment:
Let's add a comment for this parameter. The purpose is to catch any callers that do not "handle dynamification of assignment". Let's clarify what that means to handle. 

From my understanding:
1. some indexer accesses with dynamic argument get resolved statically, and they get marked with type `dynamic`
2. but assignment scenarios should adjust their indexer/operand back to their original type (non-dynamic) and instead use type `dynamic` as the result of the assignment itself
3. this flag and the assertion below help catch any new assignment scenarios and make them aware of this subtlety